File: data/Pandora.Api/V1/Controllers/Terraform-Models.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Pandora.Api.V1.Controllers;

public partial class TerraformController
{
    private class TerraformV1Response
    {
        [JsonPropertyName("dataSources")]
        public Dictionary<string, DataSourceResponse> DataSources { get; set; }

        [JsonPropertyName("resources")]
        public Dictionary<string, ResourceResponse> Resources { get; set; }
    }

    private class DataSourceResponse
    {
        // TODO: stuff and things - don't forget to account for Single vs Plural DS's

        [JsonPropertyName("apiVersion")]
        public string ApiVersion { get; set; }

        [JsonPropertyName("generate")]
        public bool Generate => GenerateModel || GenerateSchema || (PluralDetails?.Generate ?? false) || (SingularDetails?.Generate ?? false);

        [JsonPropertyName("generateModel")]
        public bool GenerateModel { get; set; }

        [JsonPropertyName("generateSchema")]
        public bool GenerateSchema { get; set; }

        [JsonPropertyName("plural")]
        public TerraformDataSourceTypeDetails? PluralDetails { get; set; }

        [JsonPropertyName("singular")]
        public TerraformDataSourceTypeDetails? SingularDetails { get; set; }

        // TODO: add other properties
    }

    private class TerraformDataSourceTypeDetails
    {
        [JsonPropertyName("description")]
        public string Description { get; set; }

        [JsonPropertyName("exampleUsageHcl")]
        public string ExampleUsageHcl { get; set; }

        [JsonPropertyName("generate")]
        public bool Generate => GenerateModel || GenerateSchema || MethodDefinition.Generate;

        [JsonPropertyName("generateModel")]
        public bool GenerateModel { get; set; }

        [JsonPropertyName("generateSchema")]
        public bool GenerateSchema { get; set; }

        [JsonPropertyName("methodDefinition")]
        public MethodDefinition MethodDefinition { get; set; }

        [JsonPropertyName("resourceLabel")]
        public string ResourceLabel { get; set; }
    }

    private class ResourceResponse
    {
        // TODO: Schema [incl. Docs], Mappings, Tests etc
        [JsonPropertyName("apiVersion")]
        public string ApiVersion { get; set; }

        [JsonPropertyName("createMethod")]
        public MethodDefinition CreateMethod { get; set; }

        [JsonPropertyName("deleteMethod")]
        public MethodDefinition DeleteMethod { get; set; }

        [JsonPropertyName("displayName")]
        public string DisplayName { get; set; }

        [JsonPropertyName("documentation")]
        public ResourceDocumentationDefinition Documentation { get; set; }

        [JsonPropertyName("generate")]
        public bool Generate => DeleteMethod.Generate || GenerateModel || GenerateIdValidation || GenerateSchema;

        [JsonPropertyName("generateModel")]
        public bool GenerateModel { get; set; }

        [JsonPropertyName("generateIdValidation")]
        public bool GenerateIdValidation { get; set; }

        [JsonPropertyName("generateSchema")]
        public bool GenerateSchema { get; set; }

        [JsonPropertyName("mappings")]
        public MappingsDefinition Mappings { get; set; }

        [JsonPropertyName("readMethod")]
        public MethodDefinition ReadMethod { get; set; }

        [JsonPropertyName("resource")]
        public string Resource { get; set; }

        [JsonPropertyName("resourceIdName")]
        public string ResourceIdName { get; set; }

        [JsonPropertyName("resourceName")]
        public string ResourceName { get; set; }

        [JsonPropertyName("schemaModelName")]
        public string SchemaModelName { get; set; }

        [JsonPropertyName("schemaModels")]
        public Dictionary<string, TerraformSchemaDefinition> SchemaModels { get; set; }

        [JsonPropertyName("tests")]
        public TerraformResourceTestsDefinition Tests { get; set; }

        [JsonPropertyName("updateMethod")]
        public MethodDefinition? UpdateMethod { get; set; }
    }

    private class ResourceDocumentationDefinition
    {
        [JsonPropertyName("category")]
        public string Category { get; set; }

        [JsonPropertyName("description")]
        public string Description { get; set; }

        [JsonPropertyName("exampleUsageHcl")]
        public string ExampleUsageHcl { get; set; }
    }

    private class MethodDefinition
    {
        [JsonPropertyName("generate")]
        public bool Generate { get; set; }

        [JsonPropertyName("methodName")]
        public string MethodName { get; set; }

        [JsonPropertyName("timeoutInMinutes")]
        public int TimeoutInMinutes { get; set; }
    }

    private class TerraformSchemaDefinition
    {
        public Dictionary<string, TerraformSchemaFieldDefinition> Fields { get; set; }
    }

    private class TerraformSchemaObjectDefinition
    {
        [JsonPropertyName("nestedObject")]
        public TerraformSchemaObjectDefinition? NestedObject { get; set; }

        [JsonPropertyName("referenceName")]
        public string? ReferenceName { get; set; }

        [JsonPropertyName("type")]
        public string Type { get; set; }
    }

    private class TerraformSchemaFieldDefinition
    {
        [JsonPropertyName("objectDefinition")]
        public TerraformSchemaObjectDefinition ObjectDefinition { get; set; }

        [JsonPropertyName("computed")]
        public bool Computed { get; set; }

        [JsonPropertyName("forceNew")]
        public bool ForceNew { get; set; }

        [JsonPropertyName("hclName")]
        public string HclName { get; set; }

        [JsonPropertyName("optional")]
        public bool Optional { get; set; }

        [JsonPropertyName("required")]
        public bool Required { get; set; }

        [JsonPropertyName("documentation")]
        public TerraformSchemaDocumentationDefinition Documentation { get; set; }

        [JsonPropertyName("validation")]
        public TerraformSchemaFieldValidationDefinition Validation { get; set; }
    }

    private class TerraformSchemaDocumentationDefinition
    {
        [JsonPropertyName("markdown")]
        public string Markdown { get; set; }
    }

    private class TerraformSchemaFieldValidationDefinition
    {
        [JsonPropertyName("type")]
        public string Type { get; set; }

        [JsonPropertyName("possibleValues")]
        public TerraformSchemaFieldValidationPossibleValuesDefinition? PossibleValues { get; set; }
    }

    private class TerraformSchemaFieldValidationPossibleValuesDefinition
    {
        [JsonPropertyName("type")]
        public string Type { get; set; }

        [JsonPropertyName("values")]
        public List<object>? Values { get; set; }
    }

    private enum TerraformSchemaFieldType
    {
        // Core items
        Boolean,
        DateTime,
        Dictionary,
        Integer,
        Float,
        List,
        Reference,
        Set,
        String,

        // CommonSchema items
        EdgeZone,
        Location,
        IdentitySystemAssigned,
        IdentitySystemAndUserAssigned,
        IdentitySystemOrUserAssigned,
        IdentityUserAssigned,
        ResourceGroup,
        Tags,
        Zone,
        Zones,
        // NOTE: this intentionally doesn't contain the Legacy Identity Types since they're normalized in the schema
        // to the regular identity types
    }

    private enum TerraformSchemaFieldValidationType
    {
        NoEmptyValue,
        FixedValues,
        // TODO: ResourceID, Range etc
    }

    private class ResourceIdMappingDefinition
    {
        [JsonPropertyName("schemaFieldName")]
        public string SchemaFieldName { get; set; }

        [JsonPropertyName("segmentName")]
        public string SegmentName { get; set; }
    }

    private class MappingsDefinition
    {
        // TODO: should we add a Generate flag to this?

        [JsonPropertyName("fields")]
        public List<FieldMappingDefinition> Fields { get; set; }

        [JsonPropertyName("modelToModel")]
        public List<ModelToModelMappingDefinition> ModelToModel { get; set; }

        [JsonPropertyName("resourceId")]
        public List<ResourceIdMappingDefinition> ResourceId { get; set; }
    }

    private class ModelToModelMappingDefinition
    {
        [JsonPropertyName("schemaModelName")]
        public string SchemaModelName { get; set; }

        [JsonPropertyName("sdkModelName")]
        public string SdkModelName { get; set; }
    }

    private class FieldMappingDefinition
    {
        [JsonPropertyName("type")]
        public string Type { get; set; }

        [JsonPropertyName("directAssignment")]
        public FieldMappingDirectAssignmentDefinition? DirectAssignment { get; set; }

        [JsonPropertyName("modelToModel")]
        public FieldMappingModelToModelDefinition? ModelToModel { get; set; }

        // TODO: BooleanEquals, Manual etc
    }

    private class FieldMappingDirectAssignmentDefinition
    {
        /// <summary>
        /// SchemaModelName is the name of the Schema Model
        /// </summary>
        [JsonPropertyName("schemaModelName")]
        public string SchemaModelName { get; set; }

        /// <summary>
        /// SchemaFieldPath is the path to the field within the SchemaModelName (e.g. `Foo` or `Foo.Bar`)
        /// </summary>
        [JsonPropertyName("schemaFieldPath")]
        public string? SchemaFieldPath { get; set; }

        /// <summary>
        /// SdkModelName is the name of the Sdk Model associated with this mapping.
        /// </summary>
        [JsonPropertyName("sdkModelName")]
        public string SdkModelName { get; set; }

        /// <summary>
        /// SdkFieldPath is the path to the field within the Sdk Model
        /// </summary>
        [JsonPropertyName("sdkFieldPath")]
        public string? SdkFieldPath { get; set; }
    }

    private class FieldMappingModelToModelDefinition
    {
        /// <summary>
        /// SchemaModelName specifies the name of the Schema Model used for this ModelToModel Mapping.
        /// </summary>
        [JsonPropertyName("schemaModelName")]
        public string SchemaModelName { get; set; }

        /// <summary>
        /// SdkModelName specifies the name of the Sdk Model used for this ModelToModel Mapping. 
        /// </summary>
        [JsonPropertyName("sdkModelName")]
        public string SdkModelName { get; set; }

        /// <summary>
        /// SdkFieldName specifies the name of the Field within the Sdk Model
        /// </summary>
        [JsonPropertyName("sdkFieldName")]
        public string SdkFieldName { get; set; }
    }

    private class TerraformResourceTestsDefinition
    {
        [JsonPropertyName("basicConfiguration")]
        public string BasicConfiguration { get; set; }

        [JsonPropertyName("requiresImportConfiguration")]
        public string RequiresImportConfiguration { get; set; }

        [JsonPropertyName("completeConfiguration")]
        public string? CompleteConfiguration { get; set; }

        [JsonPropertyName("generate")]
        public bool Generate { get; set; }

        [JsonPropertyName("otherTests")]
        public Dictionary<string, List<string>> OtherTests { get; set; }

        [JsonPropertyName("templateConfiguration")]
        public string? TemplateConfiguration { get; set; }
    }

    private enum TerraformSchemaFieldValidationPossibleValueType
    {
        Float,
        Int,
        String,
    }

    private enum TerraformSchemaFieldValidation
    {
        PossibleValues,
    }

    private enum TerraformFieldMappingType
    {
        DirectAssignment,
        ModelToModel,
        // TODO: Manual, BooleanEquals etc
    }
}

File: tools/data-api/internal/repositories/terraform_models.go
package repositories

type TerraformDetails struct {
	DataSources map[string]TerraformDataSourceDetails
	Resources   map[string]TerraformResourceDetails
}

type TerraformDataSourceDetails struct {
	Description      string
	ExampleUsageHcl  string
	Generate         bool
	GenerateModel    bool
	GenerateSchema   bool
	MethodDefinition MethodDefinition
	ResourceLabel    string
}

type TerraformResourceDetails struct {
	ApiVersion           string
	CreateMethod         MethodDefinition
	DeleteMethod         MethodDefinition
	Documentation        ResourceDocumentationDefinition
	DisplayName          string
	Generate             bool
	GenerateModel        bool
	GenerateIdValidation bool
	GenerateSchema       bool
	Label                string
	Mappings             MappingDefinition
	ReadMethod           MethodDefinition
	Resource             string
	ResourceIdName       string
	ResourceName         string
	SchemaModelName      string
	SchemaModels         map[string]TerraformSchemaModelDefinition
	Tests                TerraformResourceTestsDefinition
	UpdateMethod         *MethodDefinition
}

type MethodDefinition struct {
	Generate         bool
	MethodName       string
	TimeoutInMinutes int
}

type ResourceDocumentationDefinition struct {
	Category        string
	Description     string
	ExampleUsageHcl string
}

type MappingDefinitionType string

const (
	// DirectAssignmentTerraformFieldMappingDefinitionType specifies that this mapping defines a Direct Assignment
	// between a given Field within the Terraform Schema Model and a given Field in an SDK Model.
	DirectAssignmentTerraformFieldMappingDefinitionType MappingDefinitionType = "DirectAssignment"

	// ModelToModelTerraformFieldMappingDefinitionType specifies that this mapping defines a ModelToModel mapping
	// where the specified Schema Model should be mapped onto the given Field within a SDK Model - which relies
	// on a TerraformModelToModelMappingDefinition existing to define the mapping between the Schema Model and
	// the SDK Model.
	ModelToModelTerraformFieldMappingDefinitionType MappingDefinitionType = "ModelToModel"

	// ManualTerraformFieldMappingDefinitionType specifies that this mapping must be done manually.
	// Whilst this isn't currently used, it's piped through to allow an escape-hatch for atypical
	// scenarios requiring custom transformations.
	ManualTerraformFieldMappingDefinitionType MappingDefinitionType = "Manual"

	// TODO: support for other types of mappings e.g. BooleanEquals, BooleanInvert
)

type FieldMappingDirectAssignmentDefinition struct {
	SchemaModelName string
	SchemaFieldPath string
	SdkModelName    string
	SdkFieldPath    string
}

type FieldMappingModelToModelDefinition struct {
	SchemaModelName string
	SdkModelName    string
	SdkFieldName    string
}

type FieldManualMappingDefinition struct {
	MethodName string
}

type FieldMappingDefinition struct {
	Type             MappingDefinitionType
	DirectAssignment *FieldMappingDirectAssignmentDefinition
	ModelToModel     *FieldMappingModelToModelDefinition
	Manual           *FieldManualMappingDefinition
}

type ModelToModelMappingDefinition struct {
	SchemaModelName string
	SdkModelName    string
}

type ResourceIdMappingDefinition struct {
	SchemaFieldName    string
	SegmentName        string
	ParsedFromParentID bool
}

type MappingDefinition struct {
	Fields        []FieldMappingDefinition
	ModelToModels []ModelToModelMappingDefinition
	ResourceId    []ResourceIdMappingDefinition
}

type TerraformSchemaFieldType string

const (
	// BooleanTerraformSchemaObjectDefinitionType specifies that the Type represents a Boolean.
	BooleanTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Boolean"

	// DateTimeTerraformSchemaObjectDefinitionType specifies that the Type represents a DateTime.
	DateTimeTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "DateTime"

	// DictionaryTerraformSchemaObjectDefinitionType specifies that the Type represents a Dictionary
	// Dictionaries use a String value for a Key and the Value Type will be defined as a NestedItem
	// within the TerraformSchemaObjectDefinition.
	DictionaryTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Dictionary"

	// EdgeZoneTerraformSchemaObjectDefinitionType specifies that the Type represents an Edge Zone.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	EdgeZoneTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "EdgeZone"

	// LocationTerraformSchemaObjectDefinitionType specifies that the Type represents a Location.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	LocationTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Location"

	// FloatTerraformSchemaObjectDefinitionType specifies that the Type represents a Float.
	FloatTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Float"

	// IntegerTerraformSchemaObjectDefinitionType specifies that the Type represents an Integer.
	IntegerTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Integer"

	// ListTerraformSchemaObjectDefinitionType specifies that the Type represents a List where the Value
	// Type will be defined as a NestedItem within the TerraformSchemaObjectDefinition.
	ListTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "List"

	// ReferenceTerraformSchemaObjectDefinitionType specifies that the Type represents a Reference to
	// either a Constant or a Model.
	ReferenceTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Reference"

	// ResourceGroupTerraformSchemaObjectDefinitionType specifies that the Type represents a Resource Group Name.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	ResourceGroupTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "ResourceGroup"

	// SetTerraformSchemaObjectDefinitionType specifies that the Type is a Set where the Value
	// Type will be defined as a NestedItem within the TerraformSchemaObjectDefinition.
	SetTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Set"

	// SkuTerraformSchemaObjectDefinitionType specifies that the Type represents the name of a Sku.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	SkuTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Sku"

	// StringTerraformSchemaObjectDefinitionType specifies that the Type represents a String.
	StringTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "String"

	// SystemAssignedIdentityTerraformSchemaObjectDefinitionType specifies that the Type represents a System Assigned (Managed) Identity.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	SystemAssignedIdentityTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "SystemAssignedIdentity"

	// SystemAndUserAssignedIdentityTerraformSchemaObjectDefinitionType specifies that the Type represents a System AND User Assigned Identity
	// supporting both a System Assigned (Managed) Identity, a User Assigned Identity - AND a combination of the two used together.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	SystemAndUserAssignedIdentityTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "SystemAndUserAssignedIdentity"

	// SystemOrUserAssignedIdentityTerraformSchemaObjectDefinitionType specifies that the Type represents a System OR User Assigned Identity
	// supporting either a System Assigned (Managed) Identity or a User Assigned Identity - but not both.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	SystemOrUserAssignedIdentityTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "SystemOrUserAssignedIdentity"

	// TagsTerraformSchemaObjectDefinitionType specifies that the Type represents Tags
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	TagsTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Tags"

	// UserAssignedIdentityTerraformSchemaObjectDefinitionType specifies that the Type represents a User Assigned Identity.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	UserAssignedIdentityTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "UserAssignedIdentity"

	// ZoneTerraformSchemaObjectDefinitionType specifies that the Type represents a single (Availability) Zone.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	ZoneTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Zone"

	// ZonesTerraformSchemaObjectDefinitionType specifies that the Type represents multiple (Availability) Zones.
	// This is a CommonSchema type sourced from `hashicorp/go-azure-helpers`.
	ZonesTerraformSchemaObjectDefinitionType TerraformSchemaFieldType = "Zones"
)

type TerraformSchemaFieldObjectDefinition struct {
	NestedObject  *TerraformSchemaFieldObjectDefinition `json:"nestedObject,omitempty"`
	ReferenceName *string
	Type          TerraformSchemaFieldType
}

type TerraformSchemaDocumentationDefinition struct {
	Markdown string
}

type TerraformSchemaValidationPossibleValueType string

const (
	TerraformSchemaValidationPossibleValueTypeFloat  TerraformSchemaValidationPossibleValueType = "Float"
	TerraformSchemaValidationPossibleValueTypeInt    TerraformSchemaValidationPossibleValueType = "Int"
	TerraformSchemaValidationPossibleValueTypeString TerraformSchemaValidationPossibleValueType = "String"
)

type TerraformSchemaValidationPossibleValuesDefinition struct {
	Type   TerraformSchemaValidationPossibleValueType
	Values []interface{}
}

type TerraformSchemaValidationType string

const (
	// PossibleValuesTerraformSchemaValidationType specifies that there's a fixed set of possible values
	// allowed for this field.
	PossibleValuesTerraformSchemaValidationType TerraformSchemaValidationType = "PossibleValues"

	// TODO: we should implement `PossibleValuesFromConstant` and potentially others (NoEmptyValues/Ranges)
	// in the future
)

type TerraformSchemaValidationDefinition struct {
	Type           TerraformSchemaValidationType `json:"type"`
	PossibleValues *TerraformSchemaValidationPossibleValuesDefinition
}

type TerraformSchemaFieldDefinition struct {
	ObjectDefinition TerraformSchemaFieldObjectDefinition
	Computed         bool
	ForceNew         bool
	HclName          string
	Optional         bool
	Required         bool
	Documentation    TerraformSchemaDocumentationDefinition
	Validation       *TerraformSchemaValidationDefinition
}

type TerraformSchemaModelDefinition struct {
	Fields map[string]TerraformSchemaFieldDefinition
}

type TerraformResourceTestsDefinition struct {
	BasicConfiguration          string
	RequiresImportConfiguration string
	CompleteConfiguration       *string
	Generate                    bool
	OtherTests                  map[string][]string
	TemplateConfiguration       *string
}
